### DISCUSSION TITLE
Database Issues - Circuit module 3.4.10

### DISCUSSION THREAD

**Discussion by c-bigg** [netbox-community/netbox#12667] [Help Wanted!]
A database programming error was detected while processing this request. Common causes include the following:

Database migrations missing - When upgrading to a new NetBox release, the upgrade script must be run to apply any new database migrations. You can run migrations manually by executing python3 manage.py migrate from the command line.

Unsupported PostgreSQL version - Ensure that PostgreSQL version 10 or later is in use. You can check this by connecting to the database using NetBox's credentials and issuing a query for SELECT VERSION().

The complete exception is provided below:

<class 'django.db.utils.ProgrammingError'>

column circuits_provider.account does not exist
LINE 1: ...its_provider"."name", "circuits_provider"."slug", "circuits_...
                                                             ^
Python version: 3.8.10
NetBox version: 3.4.10

This occurs on only the circuits module.  All the other modules are working fine.

Manage.py migrate:
Applying database migrations (python3 netbox/manage.py migrate)...
Operations to perform:
  Apply all migrations: admin, auth, circuits, contenttypes, dcim, django_rq, extras, ipam, sessions, social_django, taggit, tenancy, users, virtualization, wireless
Running migrations:
  No migrations to apply.

This is from the ./upgrade.sh script and i receive the same No migrations to apply when runnning python netbox/manage.py from the virtual environment.

PostrgreSQL Version:
netbox=# SELECT version();
                                                                version                                                                 
----------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 12.14 (Ubuntu 12.14-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit
(1 row)

netbox=# \d+ circuits_provider 
                                                              Table "public.circuits_provider"
      Column       |           Type           | Collation | Nullable |                    Default                    | Storage  | Stats target | Description 
-------------------+--------------------------+-----------+----------+-----------------------------------------------+----------+--------------+-------------
 id                | bigint                   |           | not null | nextval('circuits_provider_id_seq'::regclass) | plain    |              | 
 created           | timestamp with time zone |           |          |                                               | plain    |              | 
 last_updated      | timestamp with time zone |           |          |                                               | plain    |              | 
 name              | character varying(100)   |           | not null |                                               | extended |              | 
 slug              | character varying(100)   |           | not null |                                               | extended |              | 
 comments          | text                     |           | not null |                                               | extended |              | 
 custom_field_data | jsonb                    |           | not null |                                               | extended |              | 
 description       | character varying(200)   |           | not null |                                               | extended |              | 

I dont see and account column.  How do i add it back in?  This is not a new install or upgrade, ive been using it for a while but i rarely check the circuit modules.  I am not a SQL admin (im the everything else admin) so my SQL is very rusty.  Any help would be appreciated.

**Comment by candlerb** [comment]
> column circuits_provider.account does not exist

This field was removed in Netbox 3.5 - see "Breaking changes" [here](https://github.com/netbox-community/netbox/releases/v3.5.0).

Therefore, it looks like at some point you upgraded the database to v3.5, but are now trying to run Netbox v3.4 code with it.

It's not possible to migrate the database backwards.  Either you need to run Netbox v3.5 code, or you need to restore your database from a backup made when you were running v3.4 and before you previously upgraded to v3.5.

**Reply by c-bigg** [reply]
That makes sense now.  I thought after 3.5 failed i reverted the OS VM to a previous snapshot, but apparently I didnt and i just reverted the app venv back and installed 3.4.10.  I will have to move forward to 3.5, cant go back without losing a ton of data.  Thanks for the help.

**Reply by DanSheps** [reply]
You can actually migrate backwards, but you need to do it from 3.5 code and you need to know what state **exactly** it needs to get to.  There are also some functions that do not let you roll back that might be in migrations as well (some RunPython functions but I at least try and add a reverse migration in when I do them).

@c-bigg Please mark @candlerb's answer as the correct answer and not your own.
